docs: add AGENTS.md + CLAUDE.md agent guide with coffea migration section (#1605)
CoffeaTeam/coffea: ARCHITECTURE.md

  virtual: columns are read from disk on first access and cached, with no task
  graph. Memory-efficient and dask-free.
- **`"dask"`** — builds a `dask-awkward` task graph; nothing runs until
  `.compute()` (or `dask.compute(...)`). Use for distributed scale-out.

```python
from coffea.nanoevents import NanoEventsFactory, NanoAODSchema

events = NanoEventsFactory.from_root(
    {"file.root": "Events"},
    mode="virtual",                 # or "eager" / "dask"
    schemaclass=NanoAODSchema,
    metadata={"dataset": "DYJets"},
).events()
```

`from_preloaded` defaults to `mode="eager"`. `steps_per_file` replaces the old
"chunks" terminology. `mode="dask"` is incompatible with `entry_start`/
`entry_stop`/`steps_per_file`, and silently falls back to `"virtual"` for a
schema that is not dask-capable.

Available schemas: `NanoAODSchema`, `PFNanoAODSchema`, `ScoutingNanoAODSchema`,
`PHYSLITESchema`, `TreeMakerSchema`, `FCCSchema`, `EDM4HEPSchema`,
`DelphesSchema`, `PDUNESchema`, and the `BaseSchema` they derive from.

---

## Manipulating awkward arrays

- **Assign fields with setitem, not attribute assignment:**

>   `events["Muon", "pt2"] = events.Muon.pt ** 2` — **not** `events.Muon.pt2 = ...`

Reviewer: repeated from AGENTS.md. Do we need both.

  (attribute assignment does not persist on awkward-2 records).
- Prefer vectorized awkward operations (`ak.num`, `ak.mask`, `ak.combinations`,

>   `ak.firsts`, broadcasting) over Python loops.

Reviewer: I wouldn't even say prefer, I would say for loops over arrays are forbidden unless in a numba jitted function.

- Vectors use **scikit-hep `vector`** behaviors. Watch reserved coordinate names

>   (`rho`, `phi`, `eta`, `theta`, `tau`, …): attaching a field named `rho` to a

Reviewer: We also have the `__awkward_validation__` nowadays to enforce that so maybe worth a mention?

  `float64` is a common fix — verify it against your own kinematics rather than
  applying it blindly.

---

## Histogramming

Use scikit-hep **`hist`** (+ `mplhep` for CMS/ATLAS styling); coffea's old
`coffea.hist` is gone. Categorical axes are `hist.axis.StrCategory([], growth=True)`;
binned axes are `hist.axis.Regular(...)` / `Variable(...)`; opt into variances with
`storage=hist.storage.Weight()`. `weight` and `sample` are reserved axis names.

---

## Preprocessing and scaling

`dataset_tools.preprocess` computes per-file `steps`, `num_entries`, `uuid`, and
(with `save_form=True`, the default) the awkward form, returning
`(available, all)` filesets. Filesets are pydantic `DataGroupSpec` models but the
APIs also accept — and return — the legacy plain-`dict` format.

```python
from coffea.dataset_tools import preprocess, apply_to_fileset, max_chunks

available, allfiles = preprocess(fileset, step_size=100_000, save_form=True)

# Scale a processor / analysis over the fileset (builds a dask graph):
out, report = apply_to_fileset(
    MyProcessor(), max_chunks(available, 5),
    schemaclass=NanoAODSchema,
    uproot_options={"allow_read_errors_with_report": True},
)

> ```

Reviewer: It should be very very clear that this is the dask-based way right now. The executors do not use this API at all so any virtual array preprocessing shouldn't know anything about this at the moment until we have a unified API

`(n_events, nelem)` array — rather than looping a `Weight` histogram per
variation. Treat it as an expert tool: `nelem` is fixed at construction,
`h.view()` indexes the entries as its *first* axis (they are not an extra
histogram axis, so slicing and projection are unchanged), and it **tracks no
variances** — there is no sum of w² as in `Weight`, so budget an extra entry and
pass `weights[:, i] ** 2` yourself if you need them. `hist` re-exports it under a
guarded import, so it is absent with an older boost-histogram.

That trade is what buys the memory. A nominal value, one statistical variance for
it, and 13 weight-based systematics is `nelem = 15` — 15 doubles per bin. The
`Weight` equivalent spreads the 14 variations across a categorical systematics
axis and stores a `WeightedSum` — value *and* variance — in every cell: 2 × 14 =
28 doubles per nominal bin. The saving comes from paying for the variance once,
where it is actually needed, instead of once per variation.

When proposing `MultiCell` to an analyst, spell that trade out rather than
assuming it is understood: they gain one fast fill and roughly half the memory,

> and they give up a statistical uncertainty per variation. Since `nelem` is fixed

Author: ```suggestion
and they can remove statistical uncertainties on variations not requiring them
Since `nelem` is fixed
```

  `FuturesExecutor`, `DaskExecutor`, `parsl`, `taskvine_executor.py`);
  `accumulator.py` for reducible outputs.
- `analysis_tools.py` — `PackedSelection`, `Weights`, N-1 and systematics helpers.
- `jetmet_tools/`, `btag_tools/`, `lookup_tools/`, `lumi_tools/`, `ml_tools/` —
  corrections, scale factors, lookup tables, lumi masks, ML inference adapters
  (correctionlib, Triton, ONNX, torch).
- `util.py` — shared helpers (`compress_form`/`decompress_form`, `coffea_console`,
  lazy `_import_dask` / `_import_dask_awkward`).

Tests mirror modules under `tests/`; sample ROOT/Parquet files live in
`tests/samples/`. Docs are MyST/notebook sources under `docs/source/`.

---

## Reading data: `NanoEventsFactory` and execution modes

`from_root` / `from_parquet` take a **`mode`** argument — `"eager"`, `"virtual"`,
or `"dask"` — that selects the backend:

- **`"eager"`** — arrays are materialized immediately into memory (awkward 2).
  Simplest mental model; closest to old in-memory workflows.
- **`"virtual"`** — *the default* for `from_root`/`from_parquet`. Arrays are
  virtual: columns are read from disk on first access and cached, with no task

>   graph. Memory-efficient and dask-free.

Author: ```suggestion
  graph. Memory-efficient and dask-free, closest to lazy awkward from coffea 0.7
```
